File: javascript-systempromptleakage/secure9.js
const express = require('express')
const OpenAI = require('openai')
const { FaissStore } = require('langchain/vectorstores/faiss')

const app = express()
const openai = new OpenAI()

const tools = [{ type: 'function', function: { name: 'get_data', description: 'Internal.', parameters: {} } }]

app.post('/chat', async (req, res) => {
  const vectorStore = await FaissStore.load('policy_index', null)
  const docs = await vectorStore.similaritySearch(req.body.message, 3)
  const policyText = docs.map(d => d.pageContent).join('\n')
  const response = await openai.chat.completions.create({
    model: 'gpt-4o',
    messages: [
      { role: 'system', content: policyText },
      { role: 'user', content: req.body.message },
    ],
    tools,
  })
  res.json({ reply: response.choices[0].message.content })
})

app.listen(3000)

File: go-systempromptleakage/secure9.go
package main

import (
	"encoding/json"
	"net/http"
	"strings"

	openai "github.com/sashabaranov/go-openai"
)

var client = openai.NewClient("sk-xxx")

var tools = []openai.Tool{{
	Type: openai.ToolTypeFunction,
	Function: &openai.FunctionDefinition{Name: "get_data", Description: "Internal lookup."},
}}

type retriever struct{}

func (r *retriever) GetRelevantDocuments(query string) []string {
	return []string{"Policy: refunds require approval."}
}

var ret = &retriever{}

func chatHandler(w http.ResponseWriter, r *http.Request) {
	docs := ret.GetRelevantDocuments(r.FormValue("message"))
	policyText := strings.Join(docs, "\n")
	resp, _ := client.CreateChatCompletion(r.Context(), openai.ChatCompletionRequest{
		Model: openai.GPT4o,
		Messages: []openai.ChatCompletionMessage{
			{Role: openai.ChatMessageRoleSystem, Content: policyText},
			{Role: openai.ChatMessageRoleUser, Content: r.FormValue("message")},
		},
		Tools: tools,
	})
	json.NewEncoder(w).Encode(map[string]string{"reply": resp.Choices[0].Message.Content})
}

func main() {
	http.HandleFunc("/chat", chatHandler)
	http.ListenAndServe(":8080", nil)
}
